components/webext-storage/src/api.rs

```diff
@@ -0,0 +1,481 @@
+fn get_keys(keys: JsonValue) -> Vec<(String, Option<JsonValue>)> {
+    match keys {
+        JsonValue::String(s) => vec![(s, None)],
+        JsonValue::Array(keys) => {
+            // because nothing with json is ever simple, each key may not be
+            // a string. We ignore any which aren't.
+            keys.iter()
+                .filter_map(|v| v.as_str().map(|s| (s.to_string(), None)))
+                .collect()
+        }
+        // XXX - we clone the map value here, but `remove()` doesn't need it - maybe
+        // we should take a param to indicate if the defaults are actually needed?
+        // (Or maybe lifetimes magic could make the clone unnecessary? It should have
+        // the same lifetime as `keys`)
+        JsonValue::Object(m) => m.into_iter().map(|(k, d)| (k, Some(d))).collect(),
+        _ => vec![],
+    }
+}
+
+/// The implementation of `storage[.sync].get()` - on success this always
+/// returns a Json object.
+pub fn get(conn: &Connection, ext_id: &str, keys: JsonValue) -> Result<JsonValue> {
+    // key is optional, or string or array of string or object keys
+    let maybe_existing = get_from_db(conn, ext_id)?;
+    let mut existing = match maybe_existing {
+        None => return Ok(JsonValue::Object(Map::new())),
+        Some(v) => v,
+    };
+    // take the quick path for null, where we just return the entire object.
+    if keys.is_null() {
+        return Ok(JsonValue::Object(existing));
+    }
+    // OK, so we need to build a list of keys to get.
+    let keys_and_defaults = get_keys(keys);
+    let mut result = Map::with_capacity(keys_and_defaults.len());
+    for (key, maybe_default) in keys_and_defaults {
+        // XXX - assume that if key doesn't exist, it doesn't exist in the result.
+        if let Some(v) = existing.remove(&key) {
+            result.insert(key, v);
+        } else if let Some(def) = maybe_default {
+            result.insert(key, def);
+        }
+    }
+    Ok(JsonValue::Object(result))
+}
+
+/// The implementation of `storage[.sync].remove()`. On success this returns the
+/// StorageChanges defined by the chrome API - it's assumed the caller will
+/// arrange to deliver this to observers as defined in that API.
+pub fn remove(conn: &Connection, ext_id: &str, keys: JsonValue) -> Result<StorageChanges> {
+    let mut existing = match get_from_db(conn, ext_id)? {
+        None => return Ok(StorageChanges::new()),
+        Some(v) => v,
+    };
+
+    let keys_and_defs = get_keys(keys);
+
+    let mut result = StorageChanges::with_capacity(keys_and_defs.len());
+    for (key, _) in keys_and_defs {
+        if let Some(v) = existing.remove(&key) {
+            result.push(StorageValueChange {
+                key,
+                old_value: Some(v.to_string()),
+                new_value: None,
+            });
+        }
+    }
+    if !result.is_empty() {
+        save_to_db(conn, ext_id, &JsonValue::Object(existing))?;
+    }
+    Ok(result)
```

Comment:
This "read from db, manipulate in memory, save back to the db" thing is exactly the sort of thing that I'd be worried about us not doing inside a transaction if we leave transactions up to the higher-level code. It's fine in this case because the "higher-level code" is the code of the `Repo`, which we own...but in that case couldn't we make these take a `Transaction` rather than a `Connection` in order to make the requirements more explicit?

(I have a very limited understanding of the type-level abstractions offered by rusqlite and discussed in the slack thread earlier, but this doesn't read like it's happening inside a transaction).

Reply:
I'm really on the fence here TBH. I'll make that change if there's consensus though.